### PR TITLE
Whisper pipeline: use Sampler

### PR DESCRIPTION
**Pull request by as-suvorov**
Ticket: 152889
Closes https://github.com/openvinotoolkit/openvino.genai/issues/1164

### CODE REVIEW

File: src/cpp/src/whisper_generation_config.cpp [hunk 1]
```diff
@@ -114,6 +81,15 @@ void WhisperGenerationConfig::validate() const {
         OPENVINO_ASSERT(!language.has_value(), "Cannot specify 'language' for not multilingual model.");
         OPENVINO_ASSERT(!task.has_value(), "Cannot specify 'task' for not multilingual model.");
     }
+
+    if (is_beam_search()) {
+        OPENVINO_ASSERT(num_return_sequences == 1,
```
Comment (ilya-lavrenov) [inline src/cpp/src/whisper_generation_config.cpp:86]: is it because our Whisper API does not allow to provide multiple outputs?
If yes, than even w/o beam_search we need to assert to this condition

File: src/cpp/include/openvino/genai/whisper_generation_config.hpp [hunk 2]
```diff
@@ -15,28 +16,14 @@ namespace genai {
...
-    // max_new_tokens has priority over max_length.
-    size_t max_length = SIZE_MAX;
-
-    // Whisper specific
-
     // Corresponds to the ”<|startoftranscript|>” token.
     int64_t decoder_start_token_id = 50258;
 
-    // End of stream token id.
-    int64_t eos_token_id = 50257;
```
Comment (ilya-lavrenov) [inline src/cpp/include/openvino/genai/whisper_generation_config.hpp:38]: I see that within Whisper code we use this value in some places like:
```c++
            // cannot be normal text token
            for (size_t i = 0; i < config.eos_token_id; i++) {
                logits_data[i] = -std::numeric_limits<float>::infinity();
            }
```
do we need to handle multiple stop tokens, then?

Comment (as-suvorov) [reply src/cpp/include/openvino/genai/whisper_generation_config.hpp:38]: The `eos_token_id` used here not as a stop criteria but as model characteristic. It's used to force double timestamps tokens when return timestamps enabled. HF doesn't handle multiple eos tokens in timestamps logits processor as well.

File: src/cpp/include/openvino/genai/generation_config.hpp [hunk 3]
```diff
@@ -152,7 +152,7 @@ class OPENVINO_GENAI_EXPORTS GenerationConfig {
 
     /// @brief checks that are no conflicting parameters, e.g. do_sample=true and num_beams > 1.
     /// @throws Exception if config is invalid.
-    void validate() const;
+    virtual void validate() const;
```
[the remark this answers is not in the document]
Comment (as-suvorov) [reply src/cpp/include/openvino/genai/generation_config.hpp:155]: I had a doubt about this `virtual`. I agree, removed

File: src/cpp/src/sampler.cpp [hunk 4]
```diff
@@ -549,14 +549,14 @@ std::vector<int64_t> Sampler::_try_finish_generation(SequenceGroup::Ptr & sequen
...
-        if (sampling_params.max_new_tokens <= generated_len || 
+        if (sampling_params.get_max_new_tokens(sequence_group->get_prompt_len()) <= generated_len || 
             is_stop_token_id_hit(running_sequence->get_generated_ids().back(), sampling_params.stop_token_ids) && !sampling_params.ignore_eos) {
             // stop sequence by max_new_tokens or stop token (eos included)
             running_sequence->set_status(SequenceStatus::FINISHED);
 
             if (is_stop_token_id_hit(running_sequence->get_generated_ids().back(), sampling_params.stop_token_ids) && !sampling_params.ignore_eos) {
                 running_sequence->set_finish_reason(GenerationFinishReason::STOP);
-            } else if (sampling_params.max_new_tokens == generated_len) {
+            } else if (sampling_params.get_max_new_tokens(sequence_group->get_prompt_len()) == generated_len) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:559]: sequence group knows about its generation parameters, maybe we can add a new method for sequence group `get_max_new_tokens()`, which will handle max_new_tokens internally?

File: src/cpp/src/sampler.cpp [hunk 5]
```diff
@@ -956,7 +956,7 @@ int64_t Sampler::GroupBeamSearcher::Group::finish(Beam beam, const ov::genai::Ge
     return preeempted_sequence_id;
 }
 
-void Sampler::GroupBeamSearcher::Group::is_done(const ov::genai::GenerationConfig& sampling_params) {
+void Sampler::GroupBeamSearcher::Group::is_done(const ov::genai::GenerationConfig& sampling_params, size_t prompt_length) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:959]: - `Sampler::GroupBeamSearcher::Group` holds `Beams`
- each `Beam` holds its `Sequence`
- each `Sequence` has a pointer to containing `SequenceGroup`
- `SequenceGroup` will have `get_max_new_tokens` method

Looks like we can drop all arguments from this method.

Comment (ilya-lavrenov) [reply src/cpp/src/sampler.cpp:959]: BTW, maybe we can perform such changes as a separate (follow-up) PR?
in other places we also have max_new_tokens which should be replaced new your solution.

File: src/cpp/src/whisper/models/decoder.cpp [hunk 6]
```diff
@@ -22,5 +22,55 @@ std::shared_ptr<WhisperDecoder> WhisperDecoder::from_path(const std::filesystem:
...
+                                                          const int64_t decoder_start_token_id) {
+    Tensor input_ids_tensor{ov::element::i64, {1, 1}};
+    input_ids_tensor.data<int64_t>()[0] = decoder_start_token_id;
+
+    Tensor beam_idx_tensor{ov::element::i32, {1}};
+    beam_idx_tensor.data<int32_t>()[0] = 0;
+
+    auto [output_tensor, infer_ms] = decode(encoder_hidden_state, input_ids_tensor, beam_idx_tensor);
+
+    int64_t output_token = ov::genai::utils::argmax(output_tensor, 0);
```
Comment (ilya-lavrenov) [inline src/cpp/src/whisper/models/decoder.cpp:35]: as far as I can see `argmax` is used only in Whisper now.
Should it moved to Whisper utils or similar file?

Comment (as-suvorov) [reply src/cpp/src/whisper/models/decoder.cpp:35]: Sure, moved from common utils

File: src/cpp/src/sampler.cpp (hunk 4 above)
Comment (as-suvorov) [reply src/cpp/src/sampler.cpp]: Added 'get_max_new_tokens' to sequence group

File: src/cpp/src/whisper/whisper.cpp [hunk 7]
```diff
@@ -25,6 +26,154 @@ using ov::genai::MicroSeconds;
...
+        ov::genai::do_suppress_tokens(logits, batch, config.suppress_tokens);
+
+        if (return_timestamps) {
+            const auto& generated_ids = initial_step ? std::vector<int64_t>{} : batch_to_generated_ids.at(batch);
+            ov::genai::process_whisper_timestamp_logits(logits, batch, config, generated_ids, initial_step);
+        }
+    }
+}
+
+std::pair<ov::genai::EncodedResults, bool> decode(std::shared_ptr<ov::genai::WhisperDecoder> decoder,
```
Comment (Wovchena) [inline src/cpp/src/whisper/whisper.cpp:50]:
```suggestion
std::pair<ov::genai::EncodedResults, bool> decode(const std::shared_ptr<ov::genai::WhisperDecoder>& decoder,
```

File: src/cpp/src/whisper/whisper.cpp [hunk 8]
```diff
@@ -25,6 +26,154 @@ using ov::genai::MicroSeconds;
...
+                                                  const std::vector<int64_t>& input_ids,
+                                                  const ov::Tensor& encoder_hidden_state,
+                                                  const std::shared_ptr<ov::genai::StreamerBase> streamer_ptr,
+                                                  ov::genai::Sampler& sampler,
+                                                  ov::genai::SequenceGroup::Ptr sequence_group,
```
Comment (Wovchena) [inline src/cpp/src/whisper/whisper.cpp:55]:
```suggestion
                                                  const ov::genai::SequenceGroup::Ptr& sequence_group,
```

File: src/cpp/src/whisper/whisper_utils.cpp [hunk 9]
```diff
@@ -41,6 +41,22 @@ void filter_non_segment_metrics(ov::genai::RawPerfMetrics& raw_metrics,
...
+        OPENVINO_THROW("logits batch size doesn't match the number of beams");
+    }
+
+    size_t vocab_size = logits.get_shape().back();
+    size_t batch_offset = batch_idx * logits.get_shape()[1] * vocab_size;
+    size_t sequence_offset = (logits.get_shape()[1] - 1) * vocab_size;
+    const float* logits_data = logits.data<const float>() + batch_offset + sequence_offset;
+
+    int64_t out_token = std::max_element(logits_data, logits_data + vocab_size) - logits_data;
+    float max_logit = logits_data[out_token];
```
Comment (Wovchena) [inline src/cpp/src/whisper/whisper_utils.cpp:55]:
```suggestion
```
Not used

File: src/python/py_whisper_pipeline.cpp [hunk 10]
```diff
@@ -156,6 +147,46 @@ auto whisper_generation_config_docstring = R"(
...
+        "openvino_genai.StopCriteria.EARLY", where the generation stops as soon as there are `num_beams` complete candidates;
+        "openvino_genai.StopCriteria.HEURISTIC" is applied and the generation stops when is it very unlikely to find better candidates;
+        "openvino_genai.StopCriteria.NEVER", where the beam search procedure only stops when there cannot be better candidates (canonical beam search algorithm).
+
+    Random sampling parameters:
+    temperature:        the value used to modulate token probabilities for random sampling.
+    top_p:              if set to float < 1, only the smallest set of most probable tokens with probabilities that add up to top_p or higher are kept for generation.
+    top_k:              the number of highest probability vocabulary tokens to keep for top-k-filtering.
+    do_sample:          whether or not to use multinomial random sampling that add up to `top_p` or higher are kept.
+    num_return_sequences: the number of sequences to generate from a single prompt.
```
Comment (Wovchena) [inline src/python/py_whisper_pipeline.cpp:189]: `rng_seed` is missing. Although it's not Whisper's fault because it's missing in the parent config as well. Can you add it, possible in a separate PR

**Review by Wovchena — APPROVED** [review]

File: src/python/py_whisper_pipeline.cpp (hunk 10 above)
Comment (as-suvorov) [reply src/python/py_whisper_pipeline.cpp:189]: Sure, will do
